torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -201,21 +214,26 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
       // a use with a different profiled type, we will still be correct.
       // In the future we could consider unifying the types of uses, or adding a
       // type refinement node so uses can have the correct corresponding type.
-      if (profiled_type == TensorType::get()) {
+      if (profiled_type == TensorType::get() && !input_is_optional) {
         continue;
```

Comment:
@Krovatkin curious, shouldn't this also have `it.destroyCurrent()` ? Or is there a DCE pass that will remove the profile node?